File: src/cadforge/utils/paths.py
"""Project path helpers for CadForge."""

from __future__ import annotations

import os
from pathlib import Path


def find_project_root(start: Path | None = None) -> Path | None:
    """Walk up from start directory to find project root.

    Project root is identified by the presence of CADFORGE.md
    or .cadforge/ directory.
    """
    current = (start or Path.cwd()).resolve()
    for directory in [current, *current.parents]:
        if (directory / "CADFORGE.md").exists():
            return directory
        if (directory / ".cadforge").is_dir():
            return directory
    return None


def get_project_root(start: Path | None = None) -> Path:
    """Get project root, raising if not found."""
    root = find_project_root(start)
    if root is None:
        raise FileNotFoundError(
            "No CadForge project found. Run 'cadforge init' to create one."
        )
    return root


def get_cadforge_dir(project_root: Path) -> Path:
    """Get .cadforge/ directory, creating if needed."""
    d = project_root / ".cadforge"
    d.mkdir(exist_ok=True)
    return d


def get_memory_dir(project_root: Path) -> Path:
    """Get .cadforge/memory/ directory, creating if needed."""
    d = get_cadforge_dir(project_root) / "memory"
    d.mkdir(exist_ok=True)
    return d


def get_output_dir(project_root: Path) -> Path:
    """Get output/ directory, creating if needed."""
    d = project_root / "output"
    d.mkdir(exist_ok=True)
    return d


def get_stl_output_dir(project_root: Path) -> Path:
    """Get output/stl/ directory, creating if needed."""
    d = get_output_dir(project_root) / "stl"
    d.mkdir(exist_ok=True)
    return d


def get_step_output_dir(project_root: Path) -> Path:
    """Get output/step/ directory, creating if needed."""
    d = get_output_dir(project_root) / "step"
    d.mkdir(exist_ok=True)
    return d


def get_vault_dir(project_root: Path) -> Path:
    """Get vault/ directory path."""
    return project_root / "vault"


def get_lance_dir(project_root: Path) -> Path:
    """Get .lance/ directory, creating if needed."""
    d = project_root / ".lance"
    d.mkdir(exist_ok=True)
    return d


def get_sessions_dir() -> Path:
    """Get user-level sessions directory."""
    d = Path.home() / ".cadforge" / "sessions"
    d.mkdir(parents=True, exist_ok=True)
    return d


def get_project_sessions_dir(project_root: Path) -> Path:
    """Get project-specific sessions directory under user home."""
    encoded = str(project_root).replace("/", "-").replace("\\", "-").strip("-")
    d = Path.home() / ".cadforge" / "projects" / encoded
    d.mkdir(parents=True, exist_ok=True)
    return d


def get_user_settings_path() -> Path:
    """Get user-level settings.json path."""
    return Path.home() / ".cadforge" / "settings.json"


def get_project_settings_path(project_root: Path) -> Path:
    """Get project-level settings.json path."""
    return project_root / ".cadforge" / "settings.json"


def get_user_cadforge_md() -> Path | None:
    """Get user-level CADFORGE.md if it exists."""
    p = Path.home() / ".cadforge" / "CADFORGE.md"
    return p if p.exists() else None


def get_skills_dirs(project_root: Path) -> list[Path]:
    """Get skill directories in precedence order (workspace > user > bundled)."""
    dirs = []
    workspace = project_root / "skills"
    if workspace.is_dir():
        dirs.append(workspace)
    user = Path.home() / ".cadforge" / "skills"
    if user.is_dir():
        dirs.append(user)
    bundled = Path(__file__).parent.parent / "skills" / "builtin"
    if bundled.is_dir():
        dirs.append(bundled)
    return dirs

File: packages/cli/src/config/paths.ts
/**
 * Project root discovery and path helpers.
 * Mirrors utils/paths.py from the Python codebase.
 */

import { existsSync, mkdirSync } from 'node:fs';
import { homedir } from 'node:os';
import { dirname, join, resolve } from 'node:path';

/**
 * Walk up from startDir to find the project root.
 * Project root is identified by CADFORGE.md or .cadforge/ directory.
 */
export function findProjectRoot(startDir?: string): string | null {
  let current = resolve(startDir ?? process.cwd());

  while (true) {
    if (existsSync(join(current, 'CADFORGE.md'))) return current;
    if (existsSync(join(current, '.cadforge'))) return current;
    const parent = dirname(current);
    if (parent === current) return null; // reached filesystem root
    current = parent;
  }
}

/**
 * Get project root or throw.
 */
export function getProjectRoot(startDir?: string): string {
  const root = findProjectRoot(startDir);
  if (!root) {
    throw new Error("No CadForge project found. Run 'cadforge init' to create one.");
  }
  return root;
}

/**
 * Get the .cadforge/ directory, creating it if needed.
 */
export function getCadforgeDir(projectRoot: string): string {
  const d = join(projectRoot, '.cadforge');
  mkdirSync(d, { recursive: true });
  return d;
}

/**
 * Get user-level settings.json path.
 */
export function getUserSettingsPath(): string {
  return join(homedir(), '.cadforge', 'settings.json');
}

/**
 * Get project-level settings.json path.
 */
export function getProjectSettingsPath(projectRoot: string): string {
  return join(projectRoot, '.cadforge', 'settings.json');
}

/**
 * Get user-level .cadforge directory, creating it if needed.
 */
export function getUserCadforgeDir(): string {
  const d = join(homedir(), '.cadforge');
  mkdirSync(d, { recursive: true });
  return d;
}

/**
 * Get the venv directory path for the Python engine.
 */
export function getVenvDir(): string {
  return join(getUserCadforgeDir(), '.venv');
}

